File: api/src/main/java/fr/huiitre/tools/modules/riot/sync/application/SyncValorantSkinsUseCase.java
package fr.huiitre.tools.modules.riot.sync.application;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;
import java.util.Set;
import java.util.UUID;
import java.util.stream.Collectors;

import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import fr.huiitre.tools.modules.core.module.domain.ModuleCode;
import fr.huiitre.tools.modules.core.role.domain.RoleCode;
import fr.huiitre.tools.modules.core.security.application.usecase.SecuredUseCase;
import fr.huiitre.tools.modules.riot.valorant.application.skin.view.ValorantSkinView;

@Service
@Transactional
public class SyncValorantSkinsUseCase implements SecuredUseCase {

    private final ValorantSkinDataProvider skinDataProvider;
    private final ValorantSkinSyncRepository skinSyncRepository;
    private final ValorantSkinLevelSyncRepository levelSyncRepository;
    private final ValorantSkinChromaSyncRepository chromaSyncRepository;

    public SyncValorantSkinsUseCase(
            ValorantSkinDataProvider skinDataProvider,
            ValorantSkinSyncRepository skinSyncRepository,
            ValorantSkinLevelSyncRepository levelSyncRepository,
            ValorantSkinChromaSyncRepository chromaSyncRepository) {
        this.skinDataProvider = skinDataProvider;
        this.skinSyncRepository = skinSyncRepository;
        this.levelSyncRepository = levelSyncRepository;
        this.chromaSyncRepository = chromaSyncRepository;
    }

    @Override
    public Optional<ModuleCode> requiredModule() {
        return Optional.of(ModuleCode.RIOT);
    }

    @Override
    public RoleCode requiredRole() {
        return RoleCode.TECH;
    }

    public ValorantSyncReport execute(Map<UUID, Long> weaponAssetIdToDbId) {
        List<ValorantSkinSyncData> external = skinDataProvider.fetchAll();

        Map<UUID, ValorantSkinView> currentByAssetId = skinSyncRepository.findAll().stream()
                .collect(Collectors.toMap(ValorantSkinView::assetId, it -> it));

        Set<UUID> externalAssetIds = external.stream()
                .map(ValorantSkinSyncData::getAssetId)
                .collect(Collectors.toSet());

        Map<UUID, Long> skinAssetIdToDbId = new HashMap<>();
        int created = 0;
        int updated = 0;
        int deleted = 0;

        for (ValorantSkinSyncData ext : external) {
            Long weaponId = weaponAssetIdToDbId.get(ext.getWeaponAssetId());
            ValorantSkinView existing = currentByAssetId.get(ext.getAssetId());

            if (existing == null) {
                Long newId = skinSyncRepository.save(ext, weaponId);
                skinAssetIdToDbId.put(ext.getAssetId(), newId);
                created++;
                continue;
            }

            skinAssetIdToDbId.put(ext.getAssetId(), existing.id());

            boolean changed = !Objects.equals(existing.name(), ext.getName())
                    || !Objects.equals(existing.iconUrl(), ext.getIconUrl())
                    || !Objects.equals(existing.tierUuid(), ext.getTierUuid())
                    || !Objects.equals(existing.contentTierUuid(), ext.getContentTierUuid())
                    || !Objects.equals(existing.weaponId(), weaponId);

            if (changed) {
                skinSyncRepository.update(existing.id(), ext, weaponId);
                updated++;
            }
        }

        for (ValorantSkinView current : currentByAssetId.values()) {
            if (!externalAssetIds.contains(current.assetId())) {
                skinSyncRepository.delete(current.id());
                deleted++;
            }
        }

        syncLevels(external, skinAssetIdToDbId);
        syncChromas(external, skinAssetIdToDbId);

        return new ValorantSyncReport(created, updated, deleted);
    }

    private void syncLevels(List<ValorantSkinSyncData> external, Map<UUID, Long> skinAssetIdToDbId) {
        levelSyncRepository.deleteAll();

        for (ValorantSkinSyncData skin : external) {
            Long skinId = skinAssetIdToDbId.get(skin.getAssetId());
            if (skinId == null) continue;

            for (ValorantSkinLevelSyncData level : skin.getLevels()) {
                levelSyncRepository.save(skinId, level);
            }
        }
    }

    private void syncChromas(List<ValorantSkinSyncData> external, Map<UUID, Long> skinAssetIdToDbId) {
        chromaSyncRepository.deleteAll();

        for (ValorantSkinSyncData skin : external) {
            Long skinId = skinAssetIdToDbId.get(skin.getAssetId());
            if (skinId == null) continue;

            for (ValorantSkinChromaSyncData chroma : skin.getChromas()) {
                chromaSyncRepository.save(skinId, chroma);
            }
        }
    }
}

File: api/Modules/Riot/Sync/Application/Usecases/SyncValorantSkinsUseCase.cs
using Tools.Api.Modules.Riot.Sync.Application.Ports;

namespace Tools.Api.Modules.Riot.Sync.Application.Usecases;

public sealed class SyncValorantSkinsUseCase(
    IValorantSkinDataProvider skinDataProvider,
    IValorantSkinSyncRepository skinSyncRepository,
    IValorantSkinLevelSyncRepository levelSyncRepository,
    IValorantSkinChromaSyncRepository chromaSyncRepository
)
{
    public async Task<ValorantSyncReport> Execute(Dictionary<Guid, long> weaponAssetIdToDbId)
    {
        var external = await skinDataProvider.FetchAll();

        var currentByAssetId = (await skinSyncRepository.FindAll())
            .ToDictionary(skin => skin.AssetId);

        var externalAssetIds = external.Select(skin => skin.AssetId).ToHashSet();

        var skinAssetIdToDbId = new Dictionary<Guid, long>();
        var created = 0;
        var updated = 0;
        var deleted = 0;

        foreach (var skin in external)
        {
            long? weaponId = skin.WeaponAssetId is { } weaponAssetId
                && weaponAssetIdToDbId.TryGetValue(weaponAssetId, out var resolvedWeaponId)
                    ? resolvedWeaponId
                    : null;

            if (!currentByAssetId.TryGetValue(skin.AssetId, out var existing))
            {
                skinAssetIdToDbId[skin.AssetId] = await skinSyncRepository.Save(skin, weaponId);
                created++;
                continue;
            }

            skinAssetIdToDbId[skin.AssetId] = existing.Id;

            var changed = existing.Name != skin.Name
                || existing.IconUrl != skin.IconUrl
                || existing.TierUuid != skin.TierUuid
                || existing.WeaponId != weaponId;

            if (changed)
            {
                await skinSyncRepository.Update(existing.Id, skin, weaponId);
                updated++;
            }
        }

        foreach (var current in currentByAssetId.Values.Where(skin => !externalAssetIds.Contains(skin.AssetId)))
        {
            await skinSyncRepository.Delete(current.Id);
            deleted++;
        }

        await SyncLevels(external, skinAssetIdToDbId);
        await SyncChromas(external, skinAssetIdToDbId);

        return new ValorantSyncReport(created, updated, deleted);
    }

    // Niveaux et chromas sont purgés puis réinsérés : ils n'ont pas d'identité propre à préserver,
    // et la comparaison un à un coûterait plus que la réécriture.
    private async Task SyncLevels(
        List<ValorantSkinSyncData> external,
        Dictionary<Guid, long> skinAssetIdToDbId)
    {
        await levelSyncRepository.DeleteAll();

        foreach (var skin in external)
        {
            if (!skinAssetIdToDbId.TryGetValue(skin.AssetId, out var skinId))
            {
                continue;
            }

            foreach (var level in skin.Levels)
            {
                await levelSyncRepository.Save(skinId, level);
            }
        }
    }

    private async Task SyncChromas(
        List<ValorantSkinSyncData> external,
        Dictionary<Guid, long> skinAssetIdToDbId)
    {
        await chromaSyncRepository.DeleteAll();

        foreach (var skin in external)
        {
            if (!skinAssetIdToDbId.TryGetValue(skin.AssetId, out var skinId))
            {
                continue;
            }

            foreach (var chroma in skin.Chromas)
            {
                await chromaSyncRepository.Save(skinId, chroma);
            }
        }
    }
}
